autoprotocol/informatics.py

```diff
@@ -0,0 +1,119 @@
+"""
+Contains all the Autoprotocol Informatics objects
+
+    :copyright: 2020 by The Autoprotocol Development Team, see AUTHORS
+        for more details.
+    :license: BSD, see LICENSE for more details
+
+"""
+
+from abc import abstractmethod
+
+from .compound import Compound
+from .container import Container, WellGroup
+from .util import is_valid_well
+
+
+class Informatics:
+    """
+    Base class for informatics attribute in an instruction that is to later be
+    encoded as JSON.
+    """
+
+    def __init__(self):
+        pass
+
+    @abstractmethod
+    def as_dict(self):
+        pass
+
+
+class AttachCompounds(Informatics):
+    """
+    informatics type attach_compounds is constructed as a list of dict detailing
+    new compounds
+
+    Parameters
+    ----------
+    data:
+        dict of informatics data
+    all_wells: WellGroup
+        All wells used in the instruction that are applicable for attach_compounds
+        to take effect on.
+
+    Returns
+    -------
+    AttachCompounds
+
+    Raises
+    ------
+    TypeError
+        data must be a dict
+    TypeError
+        wells must be Well, list of Well, or WellGroup
+    ValueError
+        wells must be part of wells or containers instruction operates on
+    TypeError
+        compounds must be a list
+    TypeError
+        each element in compounds must be Compound type
+    """
+
+    def __init__(self, data: dict, all_wells):
+        # turn data into specific attributes for this type of informatics
+        if not isinstance(data, dict):
+            raise TypeError(f"informatics data: {data} must be provided in a dict.")
+
+        self.wells = data["wells"]
+        self.compounds = data["compounds"]
```

Comment:
Yup, see above. This isn't as exhaustive check. If we simplify the signature, that will come naturally

Reply:
More explicitly, instead of  `def __init__(self, data: dict):`, let's do `def __init__(self, wells, compounds):`